filters/auth/grantcallback.go

```diff
@@ -0,0 +1,92 @@
+package auth
+
+import (
+	"net/http"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/filters"
+	"golang.org/x/oauth2"
+)
+
+const GrantCallbackName = "grantCallback"
+
+type grantCallbackSpec struct {
+	config OAuthConfig
+}
+
+type grantCallbackFilter struct {
+	config OAuthConfig
+}
+
+func (grantCallbackSpec) Name() string { return GrantCallbackName }
+
+func (s grantCallbackSpec) CreateFilter([]interface{}) (filters.Filter, error) {
+	return grantCallbackFilter(s), nil
+}
+
+func (f grantCallbackFilter) exchangeAccessToken(code string, redirectURI string) (*oauth2.Token, error) {
+	ctx := providerContext(f.config)
+	params := f.config.GetAuthURLParameters(redirectURI)
+	return f.config.GetConfig().Exchange(ctx, code, params...)
+}
+
+func (f grantCallbackFilter) loginCallback(ctx filters.FilterContext) {
+	req := ctx.Request()
+	q := req.URL.Query()
+
+	code := q.Get("code")
+	if code == "" {
+		badRequest(ctx)
+		return
+	}
+
+	queryState := q.Get("state")
+	if queryState == "" {
+		badRequest(ctx)
+		return
+	}
+
+	state, err := f.config.flowState.extractState(queryState)
+	if err != nil {
+		log.Errorf("Error when extracting flow state: %v.", err)
```

Comment:
`state` is under external control so maybe we should not spam log here

Reply:
Fixed